Sources/ClientRuntime/Interceptor/DefaultInterceptorContext.swift

```diff
@@ -0,0 +1,89 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+/// Default implementation for all interceptor context types.
+/// 
+/// This object will be created before operation execution, and passed through each interceptor
+/// hook in the execution pipeline.
+public class DefaultInterceptorContext<RequestType, ResponseType, AttributesType: HasAttributes>: InterceptorContext {
+    private var attributes: AttributesType
+    private var input: Any
+    private var request: RequestType?
+    private var response: ResponseType?
+    private var output: Any?
+
+    public init(input: Any, attributes: AttributesType) {
+        self.input = input
+        self.attributes = attributes
+    }
+
+    public func getInput() -> Any {
+        self.input
+    }
+
+    public func getAttributes() -> AttributesType {
+        return self.attributes
+    }
```

Comment:
Typical form for a Swift getter (and optionally setter) is a read-only or read-write property.
```
    public var attributes: AttributesType {
        return self._attributes
    }
```
(when the name of a public & private property conflict, preface the private property with an underscore.)

Reply:
I think the reason I did it this way is because the presence of certain properties on the context object differs between Interceptor hooks. For example, in `readAfterAttempt`, the service response (`response`) _may_ not be available, if we were unable to send the request, so the property is optional. Whereas in `readBeforeDeserialization`, we definitely have received a service response, so the property is non-optional. The method getters can be overloaded, so we have `getResponse() -> ResponseType?` and `getResponse() -> ResponseType` on `DefaultInterceptorContext`, but I don't know if I can do that with properties. I could add more concrete `InterceptorContext` implementations though

Reply:
Also, I realize this comment and the one below are about `input` and `attributes`, which don't have the issue I mentioned (neither does `output`), so I could also make those regular properties and leave `getRequest` + `getResponse` as methods